src/main/java/com/google/api/generator/gapic/model/RegionTag.java

```diff
@@ -0,0 +1,131 @@
+
+  public abstract String apiVersion();
+
+  public abstract String serviceName();
+
+  public abstract String rpcName();
+
+  public abstract String overloadDisambiguation();
+
+  public abstract Boolean isAsynchronous();
+
+  public static Builder builder() {
+    return new AutoValue_RegionTag.Builder()
+        .setApiVersion("")
+        .setApiShortName("")
+        .setOverloadDisambiguation("")
+        .setIsAsynchronous(false);
+  }
+
+  abstract RegionTag.Builder toBuilder();
+
+  public final RegionTag withApiVersion(String apiVersion) {
+    return toBuilder().setApiVersion(apiVersion).build();
+  }
+
+  public final RegionTag withApiShortName(String apiShortName) {
+    return toBuilder().setApiShortName(apiShortName).build();
+  }
+
+  public final RegionTag withOverloadDisambiguation(String overloadDisambiguation) {
+    return toBuilder().setOverloadDisambiguation(overloadDisambiguation).build();
+  }
+
+  @AutoValue.Builder
+  public abstract static class Builder {
+    public abstract Builder setApiVersion(String apiVersion);
+
+    public abstract Builder setApiShortName(String apiShortName);
+
+    public abstract Builder setServiceName(String serviceName);
+
+    public abstract Builder setRpcName(String rpcName);
+
+    public abstract Builder setOverloadDisambiguation(String overloadDisambiguation);
+
+    public abstract Builder setIsAsynchronous(Boolean isAsynchronous);
+
+    abstract String apiVersion();
+
+    abstract String apiShortName();
+
+    abstract String serviceName();
+
+    abstract String rpcName();
+
+    abstract String overloadDisambiguation();
+
+    abstract RegionTag autoBuild();
+
+    public final RegionTag build() {
+      setApiVersion(sanitizeAttributes(apiVersion()));
+      setApiShortName(sanitizeAttributes(apiShortName()));
+      setServiceName(sanitizeAttributes(serviceName()));
+      setRpcName(sanitizeAttributes(rpcName()));
+      setOverloadDisambiguation(sanitizeAttributes(overloadDisambiguation()));
+      return autoBuild();
+    }
+
+    private final String sanitizeAttributes(String attribute) {
+      return JavaStyle.toUpperCamelCase(attribute.replaceAll("[^a-zA-Z0-9]", ""));
+    }
+  }
+
+  public enum RegionTagRegion {
+    START,
+    END
+  }
+
+  public String generate() {
+    Preconditions.checkState(!apiShortName().isEmpty(), "apiShortName can't be empty");
+    Preconditions.checkState(!serviceName().isEmpty(), "serviceName can't be empty");
+    Preconditions.checkState(!rpcName().isEmpty(), "rpcName can't be empty");
+
+    String rt = apiShortName() + "_";
+    if (!apiVersion().isEmpty()) {
+      rt = rt + apiVersion() + "_";
+    }
+    rt = rt + "generated_" + serviceName() + "_" + rpcName();
+    if (!overloadDisambiguation().isEmpty()) {
+      rt = rt + "_" + overloadDisambiguation();
+    }
+    if (isAsynchronous()) {
+      rt = rt + "_async";
+    } else {
+      rt = rt + "_sync";
+    }
+
+    return rt.toLowerCase();
```

Comment:
Is there a requirement for converting region tag to all lowercase? I see that the format in the design doc is something like this `$automl_v1_generated_AutoMl_GetDataset_async`.

Reply:
Got it, thanks!